fast_llm/layers/transformer/config.py

```diff
@@ -177,8 +300,49 @@ class TransformerSubLayerName(str, enum.Enum):
 
 @config_class()
 class TransformerPeftConfig(PeftConfig):
+    @abc.abstractmethod
+    def apply_linear(self, linear: "LinearBase", layer_type: TransformerSubLayerName | None = None) -> "LinearLike":
+        pass
+
+    @abc.abstractmethod
+    def apply_other(self, module: "torch.nn.Module") -> "torch.nn.Module":
+        pass
+
+    @abc.abstractmethod
+    def apply_weight(self, parameter: "ParameterMeta") -> "ParameterMeta":
+        pass
+
+    @classmethod
+    def _from_dict(
+        cls,
+        default: dict[str, typing.Any],
+        strict: bool = True,
+        flat: bool = False,
+    ) -> typing.Self:
+        if cls is TransformerPeftConfig and cls.get_subclass(default.get("type")) is None:
+            # Default subclass.
+            return TransformerNoPeftConfig._from_dict(default, strict, flat)
+        return super()._from_dict(default, strict=strict, flat=flat)
+
+
+@config_class()
+class TransformerNoPeftConfig(NoPeftConfig, TransformerPeftConfig):
+    _abstract = False
+
+    def apply_linear(self, linear: "LinearBase", layer_type: TransformerSubLayerName | None = None) -> "LinearLike":
+        return super().apply_linear(linear)
+
+    def apply_other(self, module: "torch.nn.Module") -> "torch.nn.Module":
+        return module
+
+    def apply_weight(self, parameter: "ParameterMeta") -> "ParameterMeta":
+        return parameter
+
+
+@config_class()
+class TransformerLoRAConfig(LoRAConfig, TransformerPeftConfig):
     layers: list[TransformerSubLayerName] = Field(
-        default=None,
+        default=(TransformerSubLayerName.query, TransformerSubLayerName.value_),
```

Comment:
```suggestion
        default_factory=lambda: [TransformerSubLayerName.query, TransformerSubLayerName.value_],
```

Reply:
The tuple is fine, it gets transformed into a list automatically

Reply:
I'm OK with the tuple **if** there's a documented guarantee that the config layer converts it automatically and that states the conversion rules clearly.

Right now, the behaviour isn't obvious at all. The type of this field is annotated as a list but initialized as a tuple. That's a straightforward type mismatch. I wasn't aware that coercion rules existed beyond serialized data, and I doubt anyone else is either. Does it always convert tuple to list? What happens with `typing.Sequence` or other container types?

Explicit is always better than implicit. Let's document these conversions clearly, or just avoid them entirely by using explicit defaults.

Reply:
Yes, this will be part of the config documentation.